src/Features/LanguageServer/Protocol/Handler/Completion/CompletionResolveHandler.cs

```diff
@@ -122,5 +134,42 @@ private static LSP.VSCompletionItem CloneVSCompletionItem(LSP.CompletionItem com
                 Preselect = completionItem.Preselect
             };
         }
+
+        private static async Task<LSP.TextEdit> GenerateTextEditAsync(
+            Document document,
+            CompletionService completionService,
+            CompletionItem selectedItem,
+            bool snippetsSupported,
+            CancellationToken cancellationToken)
+        {
+            var documentText = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+
+            var completionChange = await completionService.GetChangeAsync(
+                document, selectedItem, cancellationToken: cancellationToken).ConfigureAwait(false);
+            var completionChangeSpan = completionChange.TextChange.Span;
+            var newText = completionChange.TextChange.NewText;
+
+            // If snippets are supported, that means we can move the caret (represented by $0) to
+            // a new location.
+            if (snippetsSupported)
+            {
+                var caretPosition = completionChange.NewPosition;
+                if (caretPosition.HasValue)
+                {
+                    // caretPosition is the absolute position of the caret in the document.
+                    // We want the position relative to the start of the snippet.
+                    var relativeCaretPosition = caretPosition.Value - completionChangeSpan.Start;
```

Comment:
technically, the caret could be placed *outside* the bounds of the text being inserted.  this would break bdly.  for now, we can check for that, and not place the caret (with a comment that this is currently unsupported).

it woudl be good to have a test for that as well.  i.e. a custom completion provider that returns data like that.

Reply:
so, it shoudl always be possible to make a test that just hardcodes in different services/providers with different behavior.  you could literally create an AllisonCompletionService htat has whatever compliant behavior we want, which we can then test this lsp code against.

Reply:
Gotcha! Just added a custom completion service that does this, thanks.